src/Features/LanguageServer/Protocol/Handler/IRequestHandler.cs

```diff
@@ -13,6 +13,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// </summary>
     internal interface IRequestHandler
     {
+        /// <summary>
+        /// Name of the LSP method to handle.
+        /// </summary>
+        string MethodName { get; }
+
+        /// <summary>
+        /// Whether or not handling this method results in changes to the current solution state.
+        /// Mutating requests will block all subsequent requests from starting until after they have
+        /// completed and mutations have been applied. See <see cref="RequestExecutionQueue"/>.
+        /// </summary>
+        bool MutatesSolutionState { get; }
+
+        /// <summary>
+        /// Whether or not the handler execution queue should build a solution that represents the LSP
+        /// state of the world. If this property is not set the handler should return null from
+        /// <see cref="IRequestHandler{RequestType, ResponseType}.GetTextDocumentIdentifier(RequestType)"/>
+        /// and <see cref="RequestContext.Solution"/> will be null.
+        /// </summary>
+        bool RequiresLSPSolution { get; }
```

Comment:
I'm trying to think of a way where we can enforce that GetTextDocumentIdentifier is null when RequiresLSPSolution is false (or at least not require the handlers to specify both separately).  

Could we potentially create two versions of `IRequestHandler<RequestType, ResponseType> : IRequestHandler`, one that has the GetTextDocumentIdentifier method (maybe IRequestHandlerWithSolution), and another that doesn't (e.g. IRequestHandler).  I think then the stateless handler could inherit from IRequestHandlerWithSolution, leaving didChange/open/close as 'special'.

Reply:
I was thinking about different abstract base classes too, mainly because I feel like `RequestsLSPSolution` will so rarely be false, its a pain to specify it everywhere, though this mornings discussion, and as you say "we should also ensure that didOpen verifies that there's a document in the workspace", means that gets a bit more problematic. We now need didOpen to not require a solution but also to specify a text document so we can validate that its found ¯\\\_(ツ)_/¯